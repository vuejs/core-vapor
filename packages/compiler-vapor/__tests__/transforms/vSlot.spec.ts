import { ErrorCodes, NodeTypes } from '@vue/compiler-core'
import {
  IRNodeTypes,
  transformChildren,
  transformElement,
  transformSlotOutlet,
  transformText,
  transformVBind,
  transformVFor,
  transformVIf,
  transformVOn,
  transformVSlot,
} from '../../src'
import { makeCompile } from './_utils'

const compileWithSlots = makeCompile({
  nodeTransforms: [
    transformText,
    transformVIf,
    transformVFor,
    transformSlotOutlet,
    transformElement,
    transformVSlot,
    transformChildren,
  ],
  directiveTransforms: {
    bind: transformVBind,
    on: transformVOn,
  },
})

describe('compiler: transform slot', () => {
  test('implicit default slot', () => {
    const { ir, code } = compileWithSlots(`<Comp><div/></Comp>`)
    expect(code).toMatchSnapshot()

    expect(ir.template).toEqual(['<div></div>'])
    expect(ir.block.operation).toMatchObject([
      {
        type: IRNodeTypes.CREATE_COMPONENT_NODE,
        id: 1,
        tag: 'Comp',
        props: [[]],
        slots: {
          default: {
            type: IRNodeTypes.BLOCK,
            dynamic: {
              children: [{ template: 0 }],
            },
          },
        },
      },
    ])
    expect(ir.block.returns).toEqual([1])
    expect(ir.block.dynamic).toMatchObject({
      children: [{ id: 1 }],
    })
  })

  test('named slots w/ implicit default slot', () => {
    const { ir, code } = compileWithSlots(
      `<Comp>
        <template #one>foo</template>bar<span/>
      </Comp>`,
    )
    expect(code).toMatchSnapshot()

    expect(ir.template).toEqual(['foo', 'bar', '<span></span>'])
    expect(ir.block.operation).toMatchObject([
      {
        type: IRNodeTypes.CREATE_COMPONENT_NODE,
        id: 4,
        tag: 'Comp',
        props: [[]],
        slots: {
          one: {
            type: IRNodeTypes.BLOCK,
            dynamic: {
              children: [{ template: 0 }],
            },
          },
          default: {
            type: IRNodeTypes.BLOCK,
            dynamic: {
              children: [{}, { template: 1 }, { template: 2 }],
            },
          },
        },
      },
    ])
  })

  test('nested slots', () => {
    const { code } = compileWithSlots(
      `<Foo>
        <template #one><Bar><div/></Bar></template>
      </Foo>`,
    )
    expect(code).toMatchSnapshot()
  })

  test('dynamic slots name', () => {
    const { ir, code } = compileWithSlots(
      `<Comp>
        <template #[name]>foo</template>
      </Comp>`,
    )
    expect(code).toMatchSnapshot()
    expect(ir.block.operation[0].type).toBe(IRNodeTypes.CREATE_COMPONENT_NODE)
    expect(ir.block.operation).toMatchObject([
      {
        type: IRNodeTypes.CREATE_COMPONENT_NODE,
        tag: 'Comp',
        slots: undefined,
        dynamicSlots: [
          {
            name: {
              type: NodeTypes.SIMPLE_EXPRESSION,
              content: 'name',
              isStatic: false,
            },
            fn: { type: IRNodeTypes.BLOCK },
          },
        ],
      },
    ])
  })

  describe('errors', () => {
    test('error on extraneous children w/ named default slot', () => {
      const onError = vi.fn()
      const source = `<Comp><template #default>foo</template>bar</Comp>`
      compileWithSlots(source, { onError })
      const index = source.indexOf('bar')
      expect(onError.mock.calls[0][0]).toMatchObject({
        code: ErrorCodes.X_V_SLOT_EXTRANEOUS_DEFAULT_SLOT_CHILDREN,
        loc: {
          start: {
            offset: index,
            line: 1,
            column: index + 1,
          },
          end: {
            offset: index + 3,
            line: 1,
            column: index + 4,
          },
        },
      })
    })

    test('error on duplicated slot names', () => {
      const onError = vi.fn()
      const source = `<Comp><template #foo></template><template #foo></template></Comp>`
      compileWithSlots(source, { onError })
      const index = source.lastIndexOf('#foo')
      expect(onError.mock.calls[0][0]).toMatchObject({
        code: ErrorCodes.X_V_SLOT_DUPLICATE_SLOT_NAMES,
        loc: {
          start: {
            offset: index,
            line: 1,
            column: index + 1,
          },
          end: {
            offset: index + 4,
            line: 1,
            column: index + 5,
          },
        },
      })
    })
  })
})
