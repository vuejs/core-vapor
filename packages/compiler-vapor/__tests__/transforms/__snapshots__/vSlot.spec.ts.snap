// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compiler: transform slot > dynamic slots name 1`] = `
"import { resolveComponent as _resolveComponent, createComponent as _createComponent, template as _template } from 'vue/vapor';
const t0 = _template("foo")

export function render(_ctx) {
  const _component_Comp = _resolveComponent("Comp")
  const n2 = _createComponent(_component_Comp, null, null, () => [{
    name: _ctx.name, 
    fn: () => {
      const n0 = t0()
      return n0
    }
  }], true)
  return n2
}"
`;

exports[`compiler: transform slot > implicit default slot 1`] = `
"import { resolveComponent as _resolveComponent, createComponent as _createComponent, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const _component_Comp = _resolveComponent("Comp")
  const n1 = _createComponent(_component_Comp, null, { default: () => {
    const n0 = t0()
    return n0
  } }, null, true)
  return n1
}"
`;

exports[`compiler: transform slot > named slots w/ implicit default slot 1`] = `
"import { resolveComponent as _resolveComponent, createComponent as _createComponent, template as _template } from 'vue/vapor';
const t0 = _template("foo")
const t1 = _template("bar")
const t2 = _template("<span></span>")

export function render(_ctx) {
  const _component_Comp = _resolveComponent("Comp")
  const n4 = _createComponent(_component_Comp, null, {
    one: () => {
      const n0 = t0()
      return n0
    }, 
    default: () => {
      const n2 = t1()
      const n3 = t2()
      return [n2, n3]
    }
  }, null, true)
  return n4
}"
`;

exports[`compiler: transform slot > nested slots 1`] = `
"import { resolveComponent as _resolveComponent, createComponent as _createComponent, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const _component_Bar = _resolveComponent("Bar")
  const _component_Foo = _resolveComponent("Foo")
  const n3 = _createComponent(_component_Foo, null, { one: () => {
    const n1 = _createComponent(_component_Bar, null, { default: () => {
      const n0 = t0()
      return n0
    } })
    return n1
  } }, null, true)
  return n3
}"
`;
